### DISCUSSION TITLE
Dockerfile-hugginface

### DISCUSSION THREAD

**Discussion by amansoniamazatic** [FlowiseAI/Flowise#1682] [General]
I want a dockerfile which clone the flowise repo and install the flowise it should be able to run on hugginface below is my code which is running properly on local machine but i am not able to run the same in hugginface if anyone has any code or idea please provide.
Note: I don't want to install flowise using npm install -g flowise i only want from github.

Below is my Dockerfile:

# Use Node.js 18 Alpine as base image
FROM node:18-alpine

# Create the /data directory with appropriate permissions
RUN mkdir -p /data && chmod -R 777 /data
USER root

# Arguments that can be passed at build time
ARG FLOWISE_PATH=/usr/local/lib/node_modules/flowise
ARG BASE_PATH=/data/.flowise
ARG DATABASE_PATH=$BASE_PATH
ARG APIKEY_PATH=$BASE_PATH
ARG SECRETKEY_PATH=$BASE_PATH
ARG LOG_PATH=$BASE_PATH/logs

# Set the ARGs as ENV
ENV FLOWISE_PATH=$FLOWISE_PATH
ENV BASE_PATH=$BASE_PATH
ENV DATABASE_PATH=$DATABASE_PATH
ENV APIKEY_PATH=$APIKEY_PATH
ENV SECRETKEY_PATH=$SECRETKEY_PATH
ENV LOG_PATH=$LOG_PATH

# Install dependencies
RUN apk add --no-cache git python3 py3-pip make g++ build-base cairo-dev pango-dev chromium

ENV PUPPETEER_SKIP_DOWNLOAD=true
ENV PUPPETEER_EXECUTABLE_PATH=/usr/bin/chromium-browser

# Clone the Flowise repository from GitHub
RUN git clone https://github.com/FlowiseAI/Flowise.git /app
WORKDIR /app

# Install Flowise dependencies
RUN yarn install
RUN yarn build

# Configure Flowise directories using the ARG
RUN mkdir -p $FLOWISE_PATH/uploads && chmod -R 777 $FLOWISE_PATH

# Expose the necessary ports (if required by Flowise)
EXPOSE 7860

# Set the entry point to run Flowise
CMD ["sh", "-c", "mkdir -p $LOG_PATH && chmod -R 777 $LOG_PATH && npx flowise start"]

**Comment by Jaredude** [comment]
What's the need/use case for using git instead of npm install -g?

**Reply by amansoniamazatic** [reply]
@Jaredude I want to try out some new functionality and don't want to run it on my local PC for 3-4 days. Because of that, I think Hugging Face is a great choice.

**Reply by Jaredude** [reply]
HF works fairly well. What's the reasoning behind not wanting to use -g on the HF Docker container though? It has no impact on your local machine anyway.

**Reply by amansoniamazatic** [reply]
@Jaredude Actually i am doing some modification in flowise forked repo to add additional functionality of using if/else directly with conversational retrieval QA chain and i able to achieve this but i want to run it as a demo in hugginface space.

**Reply by Jaredude** [reply]
AWS EC2 would be a better fit for a dev demo playground. That way, you have an instance you can ssh into and modify as needed while still having the ability to have a public facing URL. Think of the flow for when you have a code change and want to update the demo. For HG, you need to factory rebuild your space, and if something goes wrong all you have is a log file to look at (exactly the situation you're in right now). For AWS, you login to the EC2 instance, pull the latest commit, rebuild your docker image, and start the new image; and if something goes wrong, you can fix it right then.

**Reply by amansoniamazatic** [reply]
Thanks @Jaredude  for letting me know about this. In HuggingFace, I used persisted storage with MySQL by connecting it to Aiven Cloud. If any faults occur in HuggingFace or i have to factory rebuilt, there is no need to worry as I can start from where I left off. Anyway, I was able to find a way to use 'git clone' in Hugging Face.

Below is the Dockerfile code to achieve this in HuggingFace space, please let me know what your thought on this:


`# Use Node.js 18 Alpine as base image
FROM node:18-alpine

# Create the /data directory with appropriate permissions
RUN mkdir -p /data && chmod -R 777 /data

# Create the flowise directory with appropriate permissions
RUN mkdir -p /usr/local/lib/node_modules/flowise && chmod -R 777 /usr/local/lib/node_modules/flowise

# Switch to a non-root user for security
USER node

# Arguments that can be passed at build time
ARG FLOWISE_PATH=/usr/local/lib/node_modules/flowise
ARG BASE_PATH=/data/.flowise
ARG DATABASE_PATH=$BASE_PATH
ARG APIKEY_PATH=$BASE_PATH
ARG SECRETKEY_PATH=$BASE_PATH
ARG LOG_PATH=$BASE_PATH/logs
ARG PORT=7860

# Set the ARGs as ENV
ENV FLOWISE_PATH=$FLOWISE_PATH
ENV BASE_PATH=$BASE_PATH
ENV DATABASE_PATH=$DATABASE_PATH
ENV APIKEY_PATH=$APIKEY_PATH
ENV SECRETKEY_PATH=$SECRETKEY_PATH
ENV LOG_PATH=$LOG_PATH
ENV PORT=$PORT

# Install dependencies
USER root
RUN apk add --no-cache git python3 py3-pip make g++ build-base cairo-dev pango-dev chromium
USER node

ENV PUPPETEER_SKIP_DOWNLOAD=true
ENV PUPPETEER_EXECUTABLE_PATH=/usr/bin/chromium-browser

# Create the /app directory with appropriate permissions
RUN mkdir -p /home/node/app && chmod -R 777 /home/node/app
WORKDIR /home/node/app

# Clone the Flowise repository from GitHub
RUN git clone https://github.com/amansoniamazatic/Flowise.git .

# Install Flowise dependencies
RUN yarn install
RUN yarn build

# Set the entry point to run Flowise
CMD ["sh", "-c", "mkdir -p $LOG_PATH && chmod -R 777 $LOG_PATH && mkdir -p $FLOWISE_PATH/uploads && chmod -R 777 $FLOWISE_PATH/uploads && npx flowise start"]`

**Reply by Jaredude** [reply]
This updated Dockerfile is working on HF? Or it's still not working? And you mentioned that it's running fine locally?

All the items around FLOWISE_PATH are not necessary since there's no -g install

You can great rid of:
```
RUN mkdir -p /usr/local/lib/node_modules/flowise && chmod -R 777 /usr/local/lib/node_modules/flowise
ARG FLOWISE_PATH=/usr/local/lib/node_modules/flowise
ENV FLOWISE_PATH=$FLOWISE_PATH
```

You have double user switching. Seems like USER root can go above the args and env. Then everything after the install can be done as the node user.

Also, why use npx flowise start instead of yarn start?

**Reply by amansoniamazatic** [reply]
@Jaredude This Dockerfile working fine in Huggingface space also same is working fine locally. 
Also i try to remove the :
`RUN mkdir -p /usr/local/lib/node_modules/flowise && chmod -R 777 /usr/local/lib/node_modules/flowise
ARG FLOWISE_PATH=/usr/local/lib/node_modules/flowise
ENV FLOWISE_PATH=$FLOWISE_PATH`


But then getting the errors of flowise.node_modules denied access because of that i also have to switch user to root to node, Also we can use yarn start there is no particular reason to use npx flowise start.
